tensorflow/java/src/main/java/org/tensorflow/Scope.java

```diff
@@ -0,0 +1,198 @@
+   * @return a new subscope
+   * @throws IllegalArgumentException if the name is invalid
+   */
+  public Scope withSubScope(String childScopeName) {
+    return toBuilder().nameScope(nameScope.withSubScope(childScopeName)).build();
+  }
+
+  /**
+   * Return a new scope that uses the provided name for an op.
+   *
+   * <p>Operations created within this scope will have a name of the form {@code
+   * name/opName[_suffix]}. This lets you name a specific operator more meaningfully.
+   *
+   * <p>Valid operator names must match the regular expression {@code [A-Za-z0-9.][A-Za-z0-9_.\\-]*}
+   *
+   * @param opName name for an operator in the returned scope
+   * @return a new Scope that uses opName for operations.
+   * @throws IllegalArgumentException if the name is invalid
+   */
+  public Scope withOpName(String opName) {
+    return toBuilder().nameScope(nameScope.withOpName(opName)).build();
+  }
+
+  /**
+   * Create a unique name for an operator, using a provided default if necessary.
+   *
+   * <p>This is normally called only by operator building classes.
+   *
+   * <p>This method generates a unique name, appropriate for the name scope controlled by this
+   * instance. Typical operator building code might look like
+   *
+   * <pre>{@code
+   * scope.graph().opBuilder("Constant", scope.makeOpName("Constant"))...
+   * }</pre>
+   *
+   * <p><b>Note:</b> if you provide a composite operator building class (i.e, a class that adds a
+   * set of related operations to the graph by calling other operator building code) you should also
+   * create a {@link #withSubScope(String)} scope for the underlying operators to group them under a
+   * meaningful name.
+   *
+   * <pre>{@code
+   * public static Stddev create(Scope scope, ...) {
+   *   // group sub-operations under a common name
+   *   Scope group = scope.withSubScope("stddev");
+   *   ... Sqrt.create(group, Mean.create(group, ...))
+   * }
+   * }</pre>
+   *
+   * @param defaultName name for the underlying operator.
+   * @return unique name for the operator.
+   * @throws IllegalArgumentException if the default name is invalid.
+   */
+  public String makeOpName(String defaultName) {
+    return nameScope.makeOpName(defaultName);
+  }
+
+  private Scope(Builder builder) {
+    graph = builder.graph;
+    if (builder.nameScope != null) {
+      nameScope = builder.nameScope;
+    } else {
+      nameScope = NameScope.create();
+    }
+  }
+
+  private final Graph graph;
+  private final NameScope nameScope;
+
+  private Builder toBuilder() {
+    return builder(graph).nameScope(nameScope);
+  }
+
+  private static Builder builder(Graph graph) {
+    return new Builder(graph);
+  }
+
+  private static final class Builder {
+    private Builder(Graph g) {
+      graph = g;
+    }
+
+    private Builder nameScope(NameScope ns) {
+      nameScope = ns;
+      return this;
+    }
+
+    private Scope build() {
+      return new Scope(this);
```

Comment:
Appreciate the thoughtful reply. A few replies:

- RE: Builders: I took those guidelines to mean that that builders are great as a replacement for factory methods when there are many parameters, or optional parameters or when calling code is less self-documenting that it should be. However, in this particular case, the proposal is to have only a single constructor (`new Scope(graph)`) and then named methods for child scopes (`withFoo()`), hence I don't think the preconditions for those guidelines are met. The builder in this PR is private and used for the implementation and IMHO in this case they are making the implementation harder to follow. So, I'd like to lean against the use of the private builders here. Sound fair?

- RE: Helper scopes: I see. IIUC, you're trying to encapsulate all the state needed for "names" in one class, all the state needed for "control dependencies" in another class etc. So, instead of having each `withFoo()` method worry about how it handles every private state member of the `Scope` class, it can deal with groups? I can buy that - let's see how things look once the builders are gone (I suspect the implementation of both classes will become much smaller) and see how it goes then.